integration-libs/epd-visualization/components/visual-viewer/visual-viewer.service.ts

```diff
@@ -0,0 +1,1530 @@
+      // This is needed for the animation slider handle position to get updated
+      // while an animation is playing.
+      // Otherwise it typically only moves once the animation playback has paused.
+      this.changeDetectorRef.detectChanges();
+    }
+  }
+
+  private setVisualizationLoadInfo(
+    visualizationLoadInfo: VisualizationLoadInfo
+  ) {
+    this._visualizationLoadInfo = visualizationLoadInfo;
+    this.visualizationLoadInfoChange.emit(visualizationLoadInfo);
+    this.changeDetectorRef.detectChanges();
+  }
+  public get visualizationLoadInfo(): VisualizationLoadInfo {
+    return this._visualizationLoadInfo;
+  }
+  private _visualizationLoadInfo: VisualizationLoadInfo;
+  public visualizationLoadInfoChange =
+    new EventEmitter<VisualizationLoadInfo>();
+
+  public loadVisualization(
+    productCode: string
+  ): Observable<VisualizationLoadInfo> {
+    this.selectedNodeIdsSubscription?.unsubscribe();
+
+    return this.viewportAdded$.pipe(
+      mergeMap(() =>
+        this.resolveVisualization(productCode).pipe(
+          mergeMap((visualizationLoadInfo: VisualizationLoadInfo) => {
+            if (
+              visualizationLoadInfo.lookupResult ===
+              VisualizationLookupResult.UniqueMatchFound
+            ) {
+              this.sceneNodeToProductLookupService.populateMapsForScene(
+                this.sceneId
+              );
+
+              let mergedVisualizationLoadInfo: VisualizationLoadInfo = {
+                ...visualizationLoadInfo,
+                loadStatus: VisualizationLoadStatus.Loading,
+              };
+              this.setVisualizationLoadInfo(mergedVisualizationLoadInfo);
+
+              return this.loadScene(this.sceneId, this.contentType).pipe(
+                mergeMap((sceneLoadInfo: SceneLoadInfo) => {
+                  if (sceneLoadInfo.sceneLoadState === SceneLoadState.Failed) {
+                    mergedVisualizationLoadInfo = {
+                      ...visualizationLoadInfo,
+                      loadStatus: VisualizationLoadStatus.UnexpectedError,
+                      errorMessage: sceneLoadInfo.errorMessage,
+                    };
+                  } else {
+                    this.selectedNodeIdsSubscription =
+                      this.selectedNodeIds$.subscribe(
+                        this.handleSelectedNodeIds.bind(this)
+                      );
+
+                    mergedVisualizationLoadInfo = {
+                      ...visualizationLoadInfo,
+                      loadStatus: VisualizationLoadStatus.Loaded,
+                    };
+                  }
+                  this.setVisualizationLoadInfo(mergedVisualizationLoadInfo);
+                  return of(mergedVisualizationLoadInfo);
+                })
+              );
+            } else {
+              return of(visualizationLoadInfo);
+            }
+          })
+        )
+      )
+    );
+  }
+
+  private isUi5BootStrapped(): boolean {
+    return !!window.sap;
```

Comment:
I'm sorry I did QA for SSR so late... I checked now that this line throws an error in SSR, when visiting a PDP with the visualisation, e.g. (http://localhost:4000/powertools-epdvisualization-spa/en/product/CX704/7%E2%80%9Dx12%E2%80%9D%20Mini%20Metal%20Lathe). Please use `WindowRef.nativeWindow?.sap` instead.

Reply:
How to test it:
- in `server.ts` configure options `debug:true` and increase `timoeut`
```
const ngExpressEngine = NgExpressEngineDecorator.get(engine, {
  timeout: 30000,
  debug: true,
});

```
- run SSR (in spartacus repo or in the fresh customer's app installed via schematics)
- in another terminal window request for the page `curl http://localhost:4000/powertools-epdvisualization-spa/en/product/CX704/7%E2%80%9Dx12%E2%80%9D%20Mini%20Metal%20Lathe > pdp.html`

The error in the SSR process terminal window looks likes this:
```
Rendering started (/powertools-epdvisualization-spa/en/product/CX704/7%E2%80%9Dx12%E2%80%9D%20Mini%20Metal%20Lathe)
ERROR ReferenceError: window is not defined
    at VisualViewerService.isUi5BootStrapped (/Users/krzysztof/code/spartacus-latest-EPD-v6/apps/ssr/dist/ssr/server/main.js:1:5848176)
    at Observable._subscribe (/Users/krzysztof/code/spartacus-latest-EPD-v6/apps/ssr/dist/ssr/server/main.js:1:5848369)
    at Observable._trySubscribe (/Users/krzysztof/code/spartacus-latest-EPD-v6/apps/ssr/dist/ssr/server/main.js:1:4051356)
    at Observable.subscribe (/Users/krzysztof/code/spartacus-latest-EPD-v6/apps/ssr/dist/ssr/server/main.js:1:4051138)
    at MergeMapOperator.call (/Users/krzysztof/code/spartacus-latest-EPD-v6/apps/ssr/dist/ssr/server/main.js:1:4119854)
    at Observable.subscribe (/Users/krzysztof/code/spartacus-latest-EPD-v6/apps/ssr/dist/ssr/server/main.js:1:4050992)
    at MergeMapOperator.call (/Users/krzysztof/code/spartacus-latest-EPD-v6/apps/ssr/dist/ssr/server/main.js:1:4119854)
    at Observable.subscribe (/Users/krzysztof/code/spartacus-latest-EPD-v6/apps/ssr/dist/ssr/server/main.js:1:4050992)
    at MergeMapSubscriber.shareReplayOperation (/Users/krzysztof/code/spartacus-latest-EPD-v6/apps/ssr/dist/ssr/server/main.js:1:4129306)
    at Observable.subscribe (/Users/krzysztof/code/spartacus-latest-EPD-v6/apps/ssr/dist/ssr/server/main.js:1:4050992)
    ```
    
That said, the page html is successfully rendered.

Reply:
Thanks. I can reproduce this server side error. I will look into it.

Reply:
I have added some guards for the code that bootstraps and interacts with UI5.